### ISSUE TITLE
Why does npm explicitly set uid/gid on installed package files?

### ISSUE THREAD

**Issue by myndzi** [npm/npm#9032]
I spent a lot of time today running down why a sticky-bit folder's node_modules installations weren't inheriting the group id. The npm cache folder didn't affect it. The /tmp folder didn't affect it. I finally got down into the nitty gritty of the source code to find out that it apparently sets these values explicitly, rather than, I dunno, letting the operating system handle that sort of thing how it normally would otherwise.

My question is, why? It seems like extra work and less compatibility; I assume it was done for some reason or other. I have worked out a workaround using the global config file to explicitly specify a uid, but that's not really ideal since it doesn't actually help solve the original problem.

**Comment by othiym23** [comment]
When you run npm as root, it needs to explicitly set the permissions to get around the fact that `node-tar` will set the permissions to be whatever was in the tarball (and also to ensure that all the files don't end up  being owned by root). I don't think any of us know of a good reason why it needs to do this when run as an unprivileged user, so a patch that preserves this behavior when run as root, but skips it otherwise, would be welcome. That said, there are a lot of weird edge cases around permissions, and the code that does this is scattered throughout the installer (and cache) code, so it might take a few passes before everything works correctly.

**Comment by myndzi** [comment]
Yes, permissions can be very complicated, which is why I'm in favor of not messing with them at all if possible. For some context, the problem is arising for me when running npm as root under a docker host accessing a shared volume -- the files created here are owned by root, since the docker container doesn't have knowledge of other users and itself runs as root. I tried to work around this with the sticky bit but was surprised by this behavior.

It seems like the appropriate fix in this case would be to prevent `node-tar` from writing permissions according to the tar file then. Since npm is a scenario that obviously doesn't make any sense for preserving uid/gid information, the only really "compatible" fix is to let the OS do whatever its default behavior would be. If `node-tar` is what is preventing that, then that's where the fix should lie.

I don't think a conditional "do one thing or another" patch would help things in any way, here. Can you think of any other cases throughout the installer and cache code that you mention that actually have some need to explicitly set uids/gids and why that might be?

**Comment by hax** [comment]
I meet a problem which seem relate to this issue.

I want to use shared cache for npm with setting global cache directory `/data/npm-cache-dir` for all users, and use `setfacl -d -m g:users:rw /data/npm-cache-dir` to set default permissions. Normally, all new files/dirs created by users under `npm-cache-dir` will be `rwxrwxr-x`.

But `npm` behavior not like I thought. After a user install one package (and update the cache directory), other users can not install the same package because they do not have permissions. The package directory in `npm-cache-dir` is still `rwxr-xr-x` means only the original user who first install that package can install/update/clean it.

**Comment by othiym23** [comment]
Since @myndzi and I had this discussion originally, I've changed my thinking here somewhat, and now believe that npm should _ensure_ that the owner and permissions of unpacked tarballs should be set safely, and that we should change this so that packages installed as a user result in installed files owned by that user (with safe permissions), and that packages installed as root use the EUID to determine what the ownership should be set to. Also, `nobody` can _never_ be allowed to own files. As such, I've classified this as a bug, and one that the team should get to soon.

As a side note, we need a better way of dealing with Windows permissions than we currently have. This will require some R&D, because I'm not sure what tools Node natively offers for interacting with ownership and ACLs on Windows, but we should at least look at it.

**Comment by frangio** [comment]
I've just ran into a related problem with npm not respecting my permissions scheme.

I set things up so that the global installation directory is owned by root and belongs to an `npm` group. This directory has `rwx` permissions for the group and the `setgid` bit set. Additionally it has an ACL entry `default:group::rwx`, which means that contained files and directories will also have `rwx` permissions for the group. However, npm sets the permissions manually and I end up with only `rx` permissions for all subdirectories.
